cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/LogBufferService.java

```diff
@@ -0,0 +1,198 @@
+import co.cask.cdap.common.service.RetryOnStartFailureService;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.common.service.RetryStrategy;
+import co.cask.cdap.logging.framework.LogPipelineLoader;
+import co.cask.cdap.logging.framework.LogPipelineSpecification;
+import co.cask.cdap.logging.logbuffer.handler.LogBufferHandler;
+import co.cask.cdap.logging.meta.CheckpointManagerFactory;
+import co.cask.cdap.logging.pipeline.LogProcessorPipelineContext;
+import co.cask.cdap.logging.pipeline.logbuffer.LogBufferPipelineConfig;
+import co.cask.cdap.logging.pipeline.logbuffer.LogBufferProcessorPipeline;
+import co.cask.http.NettyHttpService;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.Service;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryService;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Service responsible for loading, starting and stopping log buffer pipelines, creating concurrent writer and
+ * starting netty-http service to expose endpoint to process logs.
+ *
+ * TODO CDAP-14937 add log recovery mechanism upon restart
+ */
+public class LogBufferService extends AbstractIdleService {
+  private final DiscoveryService discoveryService;
+
+  private final CConfiguration cConf;
+  private final Provider<AppenderContext> contextProvider;
+  private final CheckpointManagerFactory checkpointManagerFactory;
+  private final List<Service> pipelines = new ArrayList<>();
+
+  private Cancellable cancellable;
+  private NettyHttpService httpService;
+  private ConcurrentLogBufferWriter concurrentWriter;
+
+  @Inject
+  public LogBufferService(CConfiguration cConf, DiscoveryService discoveryService,
+                          CheckpointManagerFactory checkpointManagerFactory,
+                          Provider<AppenderContext> contextProvider) {
+    this.cConf = cConf;
+    this.contextProvider = contextProvider;
+    this.checkpointManagerFactory = checkpointManagerFactory;
+    this.discoveryService = discoveryService;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // load log pipelines
+    List<LogBufferProcessorPipeline> bufferPipelines = loadLogPipelines();
+    // start all the log pipelines
+    validateAllFutures(Iterables.transform(pipelines, Service::start));
+
+    // create concurrent writer
+    this.concurrentWriter = new ConcurrentLogBufferWriter(cConf, bufferPipelines);
+
+    // create and start http service
+    this.httpService = getHttpService();
```

Comment:
the `httpService` can be final and create it in the constructor and start it here.

Reply:
`httpService` creates `LogBufferHandler` which takes concurrentWriter. Thats why not added in the constructor.